core/src/test/java/org/jdbi/v3/core/TestHandle.java

```diff
@@ -77,4 +80,61 @@ public void unknownTransactionLevelIsOk() {
         assertThatCode(() -> h.setTransactionIsolation(TransactionIsolationLevel.UNKNOWN))
             .doesNotThrowAnyException();
     }
+
+    @Test
+    public void testAutocommitFailDoesntLeak() {
+        final BoomHandler handler = new BoomHandler();
+        dbRule.getJdbi().setTransactionHandler(handler);
+        final Handle h = dbRule.openHandle();
+
+        assertThat(h.isClosed()).isFalse();
+
+        handler.failTest = true;
+        assertThatThrownBy(h::close)
+            .isInstanceOf(CloseException.class);
+
+        assertThat(h.isClosed()).isTrue();
+    }
+
+    @Test
+    public void testRollbackFailDoesntLeak() throws Exception {
+        final BoomHandler handler = new BoomHandler();
+        dbRule.getJdbi().setTransactionHandler(handler);
+        final Handle h = dbRule.openHandle();
+
+        assertThat(h.isClosed()).isFalse();
+
+        handler.failRollback = true;
+        assertThatThrownBy(() -> h.useTransaction(h2 -> h2.execute("insert into true")))
+            .isInstanceOf(UnableToCreateStatementException.class);
+        assertThat(h.isInTransaction()).isTrue(); // rollback failed!
```

Comment:
Comments clarifying tests should be added with `.describedAs(String)` instead of java source comments